### ISSUE TITLE
twisted.intenet.stdio does not read console input on Windows

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2135]
|<img alt="synapsis's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| synapsis reported|
|-|-|
|Trac ID|trac#2135|
|Type|defect|
|Created|2006-09-29 17:04:16Z|

I have tried to run the stdin and stdiodemo on Windows XP SP2 but what I type from my keyboard is not read.

I'm using Twisted 2.4, but I have tried also with the latest versions of _win32stdio and _pollingfile.

Moreover I do not see the cursor/caret.

<details><summary>Searchable metadata</summary>

```
trac-id__2135 2135
type__defect defect
reporter__synapsis synapsis
priority__lowest lowest
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__duplicate duplicate
component__core core
keywords__ 
time__1159549456000000 1159549456000000
changetime__1238625554000000 1238625554000000
version__None None
owner__ 
cc__manlio_perillo@... cc__teratorn
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="synapsis's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_1"></a>|synapsis commented|
|-|-|

Here is the problem:

Traceback (most recent call last):
  File "stdiodemo.py", line 74, in ?
    reactor.run()
  File "C:\Python2.4\lib\site-packages\twisted\internet\posixbase.py", line 218,
 in run
    self.mainLoop()
  File "C:\Python2.4\lib\site-packages\twisted\internet\posixbase.py", line 226,
 in mainLoop
    self.runUntilCurrent()
--- <exception caught here> ---
  File "C:\Python2.4\lib\site-packages\twisted\internet\base.py", line 555, in r
unUntilCurrent
    call.func(*call.args, **call.kw)
  File "_pollingfile.py", line 72, in _pollEvent
    workUnits += resource.checkWork()
  File "_pollingfile.py", line 114, in checkWork
    buffer, bytesToRead, result = win32pipe.PeekNamedPipe(self.pipe, 1)
pywintypes.error: (6, 'PeekNamedPipe', 'Handle non valido.')


Note: I have obtained this traceback by simply re raising the exception in checkWork.
      'non valido' means 'not valid'

P.S.:
In the msdn documentation it not documented that values returned by GetStdHandle can be used with pipe functions.

Why not to use PeekConsoleInput?

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_2"></a>|@glyph commented|
|-|-|

I very nearly resolved this as invalid, but there is _maybe_ a valid use-case here... Twisted programs can read and write stdio on UNIX are almost usable interactively, and the demos mentioned here imply that they are.

However, ```twisted.internet.stdio``` is not designed for console I/O on Windows, and for more advanced uses (```twisted.conch.stdio```, for example) it is infeasible that Windows could ever be made to work properly.  So, this is much more of a feature request than a bug report.

If you want console I/O that works properly, consider switching operating systems.  Windows is, and will always be, utter garbage in this respect.

**Comment by twisted-trac** [comment]
|<img alt="synapsis's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_3"></a>|synapsis commented|
|-|-|

Replying to [glyph](#note_2):
> I very nearly resolved this as invalid, but there is _maybe_ a valid use-case here... Twisted programs can read and write stdio on UNIX are almost usable interactively, and the demos mentioned here imply that they are.
>
> However, ```twisted.internet.stdio``` is not designed for console I/O on Windows, and for more advanced uses (```twisted.conch.stdio```, for example) it is infeasible that Windows could ever be made to work properly.  So, this is much more of a feature request than a bug report.
>

Yes.
See ticket [#2157](https://github.com/twisted/twisted/issues/2157).

> If you want console I/O that works properly, consider switching operating systems.  Windows is, and will always be, utter garbage in this respect.

This is not the right answer.
It is possible to add async console I/O un Windows.

**Comment by twisted-trac** [comment]
|<img alt="czzl's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_5"></a>|czzl set status to `closed`|
|-|-|

Closed as duplicate of [#2157](https://github.com/twisted/twisted/issues/2157).

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
